### PR TITLE
fix(qwen35): write prefill K/V directly to paged cache (#252)

### LINKED ISSUE #252: qwen35: prefill stages the whole sequence in a transient buffer (large allocation + hard length cap)
### Background

Qwen3.5-4B prefill materializes the full sequence's K/V in a contiguous HND staging buffer before copying into the paged pool, with the staging size fixed by a hard `MAX_SEQ` constant.

### Current state

- Each long-prompt prefill transiently allocates on the order of hundreds of MB, then pays an extra device-to-device pass to scatter into pages.
- Prompts beyond the constant cap cannot be served regardless of available KV memory; the cap is far below the model's configured max positions.
- The full-attention layers already read paged KV at decode — the staging is a prefill-only detour.

### Desired outcome

- Prefill writes K/V directly into paged blocks; the transient buffer and the hard cap disappear.
- Admissible prompt length becomes a function of KV pool capacity, consistent with the RoPE and admission fixes on this roadmap.

### Acceptance boundary

- Greedy parity on existing e2e baselines (or the logits gate, once landed) — this is a memory-layout change, not a numeric one.
- Peak prefill memory measured before/after for a long prompt.

### PR DESCRIPTION
**Pull request by wjinxu**
## What & why

Fixes #252. Qwen3.5 full-attention prefill staged the whole sequence in a transient HND buffer, then scattered it into the paged KV cache:  prep kernel → HND staging buffer → scatter kernel → paged KV cache

The HND buffer (`KVCache`) is allocated once and held for the whole runtime, sized for the worst-case `MAX_SEQ` (20000) regardless of the actual prompt — a large, permanent allocation for a purely transient purpose, and the source of the hard length cap.

## Change

Write K/V **directly into the paged KV cache** from the prep kernels, removing the staging buffer and the scatter step: paged-prep kernel → paged KV cache

- New `prefill_qk_norm_rope_hd256_paged_kernel` / `prefill_v_cache_write_hd256_paged_kernel`
  (take the paged block table, write in place).
- Drop the full-attention scatter (`paged_kv_scatter_cuda`) from prefill.
- Remove the `MAX_SEQ` cap; prompt length is now bounded by `max_position_embeddings`.
- Delete the now-orphaned non-paged prep kernels + FFI.

## Evidence (nsys, 8192-token prefill)

| | baseline | this PR |
|---|---|---|
| `AppendPagedKVCacheKernel` (scatter) | 8 calls | gone |
| HND staging allocations | 16 × 39 MiB = 625 MiB | 0 |
| compute kernels (gemm / gdr / flashinfer attn) | — | unchanged (<2%) |

Pure memory win, no compute regression.

## Validation
- `hf_golden_gate` (--test-threads=1): logits match HF within tolerance, short + long (8192).
- `e2e_scheduler`: scheduler liveness OK.
- 20001-token smoke prefill (previously rejected by MAX_SEQ) now runs.

### CODE REVIEW

File: pegainfer-qwen35-4b/src/prefill.rs [hunk 1]
```diff
@@ -220,15 +220,18 @@ impl Qwen35Model {
         let mut attn_out_batch = HiddenStates::zeros(&self.ctx, attn_out_dim, seq_len)?;
 
         let base_pos = kv_cache.len();
```
[the remark this answers is not in the document]
Comment (wjinxu) [reply pegainfer-qwen35-4b/src/prefill.rs]: fixed

File: pegainfer-qwen35-4b/src/prefill.rs [hunk 2]
```diff
@@ -26,9 +29,8 @@ fn checked_prefill_end_pos(
         anyhow::anyhow!("Qwen3.5 prefill position overflow: base_pos={base_pos}, seq_len={seq_len}")
     })?;
     anyhow::ensure!(
-        end_pos <= MAX_SEQ,
-        "Qwen3.5 prefill HND staging supports end_pos <= {MAX_SEQ}, requested end_pos={end_pos}; max_position_embeddings={}",
-        max_position_embeddings
+        end_pos <= max_position_embeddings,
```
[the remark this answers is not in the document]
Comment (wjinxu) [reply pegainfer-qwen35-4b/src/prefill.rs:31]: fixed

**Comment by xiaguan** [comment]
Mostly LGTM. The direct-paged rewrite checks out where it matters: the write offset matches the NHD layout the FlashInfer read path expects, `base_pos` is correctly recovered from `kv_state` after the `advance`, `page_indices[pos / page_size]` resolves the right physical page across page boundaries, and the `seq_len → grid.x` swap is the correct fix for the >65535 case. Dropping the scatter + ~625 MiB staging is a clean win.

A few things, none blocking:

1. (nit) The `start_pos_ptr` comment still reads "GPU-resident for CUDA Graph safety", but prefill isn't graph-captured and `start_pos_cpu` is freshly `clone_htod`'d each call — that rationale is stale carry-over from decode. Worth dropping/correcting inline.

2. On the `base_pos` fix: I don't think a test is worth adding here. A UT asserting `seq_len() - seq_len` would just restate the arithmetic, and the fix only bites once `base_pos > 0` is reachable — which the current serving path never produces (`batch_prefill` prefills each request exactly once, base_pos == 0). Rather than test a path nothing calls today, could you open a sub-issue to add a `base_pos > 0` prefill regression test alongside chunked/prefix-reuse prefill, whenever that lands? That's when an E2E actually has teeth.

3. (follow-up, not this PR) Removing `MAX_SEQ` also drops the cap that incidentally bounded the prefill *scratch*. `GdrChunkwiseScratch35::estimate_bytes` scales linearly with seq_len, but `weights.rs` still reserves only `SCRATCH_ESTIMATE_SEQ = 20_000`, and the scheduler's admission bounds KV pages, not scratch. So a prompt that fits the KV pool but runs well past ~20k tokens can now OOM mid-prefill inside `GdrChunkwiseScratch35::new` — after `kv_state.advance`, leaving it advanced on the error path — instead of being rejected up front. #252 explicitly defers memory-bounded admission to the roadmap, so this isn't for this PR; mind opening a sub-issue under the admission workstream so it doesn't get lost? (Either a scratch-honest backstop in `checked_prefill_end_pos`, or folding scratch into admission.)

**Comment by wjinxu** [comment]
Thanks for the thorough review!

1. (nit) Fixed in 59eb171 — dropped the stale "CUDA Graph safety" rationale on `start_pos_ptr` in both paged prefill kernels; it now just describes the device pointer to this prefill's base position.

2. (base_pos test) Agreed — a UT would only restate the arithmetic, and `base_pos > 0` isn't reachable on today's serving path. Opened #313 to add a `base_pos > 0` prefill regression test alongside chunked / prefix-reuse prefill, when that lands.

3. (prefill scratch admission) removing `MAX_SEQ` also dropped the incidental bound on prefill scratch, so a prompt that fits the KV pool but exceeds ~20k tokens can OOM inside `GdrChunkwiseScratch35::new` (after `kv_state.advance`). Opened #314  under the admission workstream to either add a scratch-honest backstop in `checked_prefill_end_pos` or fold scratch into admission.
